web/src/components/storage/DevicesManager.js

```diff
@@ -0,0 +1,215 @@
+   * @param {Action[]} actions - Actions to perform from system to staging.
+   */
+  constructor(system, staging, actions) {
+    this.system = system;
+    this.staging = staging;
+    this.actions = actions;
+  }
+
+  /**
+   * System device with the given SID.
+   *
+   * @param {Number} sid
+   * @returns {StorageDevice|undefined}
+   */
+  systemDevice(sid) {
+    return this.#device(sid, this.system);
+  }
+
+  /**
+   * Staging device with the given SID.
+   *
+   * @param {Number} sid
+   * @returns {StorageDevice|undefined}
+   */
+  stagingDevice(sid) {
+    return this.#device(sid, this.staging);
+  }
+
+  /**
+   * Whether the given device exists in system.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  existInSystem(device) {
+    return this.#exist(device, this.system);
+  }
+
+  /**
+   * Whether the given device exists in staging.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  existInStaging(device) {
+    return this.#exist(device, this.staging);
+  }
+
+  /**
+   * Whether the given device is going to be formatted.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  hasNewFilesystem(device) {
+    if (!device.filesystem) return false;
+
+    const systemDevice = this.systemDevice(device.sid);
+    const systemFilesystemSID = systemDevice?.filesystem?.sid;
+
+    return device.filesystem.sid !== systemFilesystemSID;
+  }
+
+  /**
+   * Whether the given device is going to be shrunk.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  isShrunk(device) {
+    return this.shrinkSize(device) > 0;
+  }
+
+  /**
+   * Amount of bytes the given device is going to be shrunk.
+   *
+   * @param {StorageDevice} device
+   * @returns {Number}
+   */
+  shrinkSize(device) {
+    const systemDevice = this.systemDevice(device.sid);
+    const stagingDevice = this.stagingDevice(device.sid);
+
+    if (!systemDevice || !stagingDevice) return 0;
+
+    const amount = systemDevice.size - stagingDevice.size;
+    return amount > 0 ? amount : 0;
+  }
+
+  /**
+   * Disk devices and LVM volume groups used for the installation.
+   *
+   * @note The used devices are extracted from the actions.
+   *
+   * @returns {StorageDevice[]}
+   */
+  usedDevices() {
+    const isTarget = (device) => device.isDrive || device.type === "lvmVg";
+
+    // Check in system devices to detect removals.
+    const targetSystem = this.system.filter(isTarget);
+    const targetStaging = this.staging.filter(isTarget);
+
+    const sids = targetSystem.concat(targetStaging)
+      .filter(d => this.#isUsed(d))
+      .map(d => d.sid);
+
+    return compact(uniq(sids).map(sid => this.stagingDevice(sid)));
+  }
+
+  /**
+   * Devices deleted.
+   *
+   * @note The devices are extracted from the actions.
+   *
+   * @returns {StorageDevice[]}
+   */
+  deletedDevices() {
+    return this.#deleteActionsDevice().filter(d => !d.isDrive);
```

Comment:
I'm not sure this is the most general fix for the problem it tries to address.

I would say it makes more sense (although likely is most complex) to get the systems from the disk and filter out the subdevices (partitions) of that disk.

Reply:
Note that `#deletedDevices` is not used for gettng the deleted systems. This method is for getting the list of deleted devices (partitions, VGs, etc). And it is filtering drives out to not include them as deleted (i.e., when a partition table is deleted or a file system is removed from a direclty formatted drive). Maybe we could simplify the detection of deleted devices by simply returning the system devices that do not appear in staging (instead of find them in the actions).

Then, for the list of systems, the strategy is similar to what you suggested, but the other way around. That is, the devices with partition table are not considered (because the systems come from the partitions), see `#deletedSystems`.

Reply:
Discussed offline. Let's leave it as it is for now.